subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/DefaultBuildCacheController.java

```diff
@@ -93,89 +117,103 @@ public boolean isEmitDebugLogging() {
-        final Unpack<T> unpack = new Unpack<>(command);
-
+    public Optional<BuildCacheLoadMetadata> load(BuildCacheKey key, CacheableEntity entity) {
+        Optional<BuildCacheLoadMetadata> result = Optional.empty();
         if (local.canLoad()) {
-            try {
-                local.load(command.getKey(), unpack);
-            } catch (Exception e) {
-                throw new GradleException("Build cache entry " + command.getKey().getHashCode() + " from local build cache is invalid", e);
-            }
-
-            if (unpack.result != null) {
-                return Optional.of(unpack.result.getMetadata());
-            }
+            result = loadLocal(key, entity);
         }
-
-        if (remote.canLoad()) {
-            tmp.withTempFile(command.getKey(), file -> {
-                LoadTarget loadTarget = new LoadTarget(file);
-                remote.load(command.getKey(), loadTarget);
-
-                if (loadTarget.isLoaded()) {
-                    try {
-                        unpack.execute(file);
-                    } catch (Exception e) {
-                        throw new GradleException("Build cache entry " + command.getKey().getHashCode() + " from remote build cache is invalid", e);
-                    }
-                    if (local.canStore()) {
-                        local.store(command.getKey(), file);
-                    }
-                }
-            });
+        if (!result.isPresent() && remote.canLoad()) {
+            result = loadRemoteAndStoreResultLocally(key, entity);
         }
+        return result;
+    }
 
-        BuildCacheLoadCommand.Result<T> result = unpack.result;
-        return result == null
-            ? Optional.empty()
-            : Optional.of(result.getMetadata());
+    private Optional<BuildCacheLoadMetadata> loadLocal(BuildCacheKey key, CacheableEntity entity) {
+        try {
+            return local.load(key, file -> unpack(key, entity, file));
+        } catch (Exception e) {
+            throw new GradleException("Build cache entry " + key.getHashCode() + " from local build cache is invalid", e);
+        }
     }
 
-    private class Unpack<T> implements Action<File> {
-        private final BuildCacheLoadCommand<T> command;
+    private Optional<BuildCacheLoadMetadata> loadRemoteAndStoreResultLocally(BuildCacheKey key, CacheableEntity entity) {
+        AtomicReference<Optional<BuildCacheLoadMetadata>> result = new AtomicReference<>(Optional.empty());
+        tmp.withTempFile(key, file -> {
+            if (remote.load(key, file)) {
+                try {
+                    result.set(Optional.of(unpack(key, entity, file)));
+                } catch (Exception e) {
+                    throw new GradleException("Build cache entry " + key.getHashCode() + " from remote build cache is invalid", e);
+                }
+            }
+            if (local.canStore()) {
+                local.store(key, file);
+            }
```

Comment:
I guess we could remove `canStore()` as well, and simply have `local.store()` do its thing if it can.

Reply:
I moved these checks inside `store/load` methods, but in one case we still need `canStore` method so we don't create a temp file. Let me know what you think.